### PR TITLE
Evolutionary algorithm for structure learning, builders decomposing and refactoring

### PR DESCRIPTION
**Pull request by jrzkaminski**
Here is the list of what's done:

- Builders are decomposed, now they are separated in different files
- Evolutionary algorithm added, covered with tests and described in documentation
- K2 score is now default score function for HC and a default metric for evolutionary algorithm
- SHD and F1 calculation function added to MathUtils, initially needed for tests, but can be used by users
- Added EvoUtils file that holds constraints, mutations, metrics for evolutionary algorithm
- GOLEM framework is added as a dependency
- autopep8 changes to some other files

There are still a couple of tweaks left to do, but general structure, classes and so on can already be reviewed

### CODE REVIEW

File: bamt/builders/builders_base.py [hunk 1]
```diff
@@ -111,6 +107,7 @@ def get_family(self):
 
         ordered = gru.toporder(self.skeleton['V'], self.skeleton['E'])
         notOrdered = [node.name for node in self.skeleton['V']]
+        notOrdered = [node.name for node in self.skeleton['V']]
```
Comment (Roman223) [inline bamt/builders/builders_base.py:110]: Repeated row?

File: bamt/builders/evo_builder.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            objective, data=data)
+
+        # Run the optimization
+        optimized_graph = optimizer.optimise(objective_eval)[0]
+
+        # Get the best graph
+        best_graph_edge_list = optimized_graph.operator.get_edges()
+        best_graph_edge_list = self._convert_to_strings(best_graph_edge_list)
+
+        print('Best graph: ', best_graph_edge_list)
```
Comment (Roman223) [inline bamt/builders/evo_builder.py:182]: I think, the better way here is to save it somewhere, or at least add a flag to control verbosity

Comment (jrzkaminski) [reply bamt/builders/evo_builder.py:182]: It was temporary of debugging, I'll remove it

File: bamt/builders/builders_base.py (hunk 1 above)
Comment (jrzkaminski) [reply bamt/builders/builders_base.py:110]: Did not notice, thanks!

File: bamt/builders/evo_builder.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                       'init_nodes': None,
+                       'remove_init_edges': True,
+                       'white_list': None,
+                       'bl_add': None}
+        self.default_n_jobs = -1
+        self.default_pop_size = 15
+        self.default_crossover_prob = 0.9
+        self.default_mutation_prob = 0.8
+        self.default_max_arity = 10
+        self.default_max_depth = 10
```
Comment (KarinaKarina6) [inline bamt/builders/evo_builder.py:72]: We don't have to limit the graph. Just in case, set the value to 100

File: bamt/builders/evo_builder.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            evo.CustomGraphModel(
+                nodes=kwargs.get(
+                    'init_nodes', [
+                        evo.CustomGraphNode(node_type) for node_type in nodes_types]))]
+
+        # Define the requirements for the evolutionary algorithm
+        requirements = GraphRequirements(
+            max_arity=kwargs.get(
+                'max_arity', self.default_max_arity), max_depth=kwargs.get(
+                'timeout', self.default_max_depth),
```
Comment (KarinaKarina6) [inline bamt/builders/evo_builder.py:121]: There should be a "max_depth" key here

File: bamt/builders/evo_builder.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            timeout=timedelta(
+                seconds=kwargs.get(
+                    'timeout', self.default_timeout)),
```
Comment (KarinaKarina6) [inline bamt/builders/evo_builder.py:124]: I think the time should be in minutes

File: bamt/builders/evo_builder.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        # Set the constraints for the graph
+
+        if kwargs.get('blacklist', None) is not None:
+            self.default_constraints.append(evo.has_no_blacklist_edges)
+        if kwargs.get('whitelist', None) is not None:
+            self.default_constraints.append(evo.has_only_whitelist_edges)
+
+        constraints = kwargs.get(
+            'custom_constraints',
+            self.default_constraints)
```
Comment (KarinaKarina6) [inline bamt/builders/evo_builder.py:152]: If a person sets custom_constraints, then mandatory constraints must also be added:
             [has_no_self_cycled_nodes,
             has_no_cycle,
             evo.has_no_duplicates]

File: bamt/utils/EvoUtils.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+from golem.core.dag.convert import graph_structure_as_nx_graph
+from golem.core.dag.graph_utils import ordered_subnodes_hierarchy
+from golem.core.optimisers.graph import OptGraph, OptNode
+
+
+class CustomGraphModel(OptGraph):
+    def evaluate(self, data: pd.DataFrame):
+        nodes = data.columns.to_list()
+        _, labels = graph_structure_as_nx_graph(self)
+        return len(nodes)
```
Comment (KarinaKarina6) [inline bamt/utils/EvoUtils.py:16]: Why do we need this method?

File: bamt/builders/evo_builder.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            descriptor: Dict[str, Dict[str, str]],
+            regressor: Optional[object],
+            has_logit: bool,
+            use_mixture: bool):
+        super(
+            EvoStructureBuilder,
+            self).__init__(
+            data=data,
+            descriptor=descriptor,
+            regressor=regressor)
```
Comment (KarinaKarina6) [inline bamt/builders/evo_builder.py:56]: Why in "build" we can specify both "classifier" and "regressor" and in "EvoStructureBuilder" only "regressor"?

File: docs/source/models/bayesiannetworks/base_network.rst [hunk 9]
```diff
@@ -12,4 +15,64 @@ The ``DiscreteBN`` and ``ContinuousBN`` are two BN types that are used to repres
...
+        :type data: DataFrame
+        :param classifier: A classification model for discrete nodes, defaults to None.
+        :type classifier: Optional[object]
+        :param regressor: A regression model for continuous nodes, defaults to None.
+        :type regressor: Optional[object]
+
+        :Keyword Args:
+            * *init_nodes* (list) -- Initial nodes to be included in the population.
+            * *max_arity* (int) -- Maximum arity for the evolutionary algorithm.
+            * *timeout* (int) -- Timeout for the evolutionary algorithm in seconds.
```
Comment (KarinaKarina6) [inline docs/source/models/bayesiannetworks/base_network.rst:64]: Also minutes

File: bamt/builders/evo_builder.py (hunk 8 above)
Comment (jrzkaminski) [reply bamt/builders/evo_builder.py:56]: Classifier is already there, it is being taken from ``build`` function

File: bamt/utils/EvoUtils.py (hunk 7 above)
Comment (jrzkaminski) [reply bamt/utils/EvoUtils.py:16]: not sure, but it was there in  GOLEM examples, we might need it in the future

**Review by KarinaKarina6 — APPROVED** [review]

**Review by Roman223 — APPROVED** [review]

**Review by Anaxagor — APPROVED** [review]

**Pull request merged**